bedms/utils_train.py

```diff
@@ -0,0 +1,692 @@
+
+
+def load_and_preprocess(file_path: str) -> pd.DataFrame:
+    """
+    Loads and Preprocesses each csv file as a Pandas DataFrame.
+
+    :param str file_path: Path to each csv file.
+    :return pandas.DataFrame: df of each csv file.
+    """
+    df = pd.read_csv(file_path, sep=",")
+    df.replace("NA", np.nan, inplace=True)
+    for column in df.columns:
+        most_common_val = df[column].mode().iloc[0]
+        df[column] = df[column].fillna(most_common_val)
+    return df
+
+
+def accumulate_data(
+    files: List[Tuple[str, str]]
+) -> Tuple[List[List[List[str]]], List[List[List[str]]], List[pd.Index]]:
+    """
+    Accumulates data from multiple files into lists.
+
+    :param List[Tuple[str, str]] files: List containing
+        sublists of values or header files.
+    :return Tuple[List[List[List[str]]], List[List[List[str]]],[List[str]]:
+        Lists of values, headers, labels.
+    A tuple containing three lists:
+        - A nested list of values (list of tables where
+            each table is a list of lists for columns),
+        - A nested list of headers (similar structure to values),
+        - A list of Pandas Index objects containing column labels.
+    """
+    x_values_list = []
+    x_headers_list = []
+    y_list = []
+    for values_file, headers_file in files:
+        df_values = load_and_preprocess(values_file)
+        df_headers = load_and_preprocess(headers_file)
+        df_values = df_values.fillna("")
+        df_headers = df_headers.fillna("")
+        y = df_values.columns
+        table_list = []
+        # values list
+        for col in df_values.columns:
+            sublist_list = df_values[col].tolist()
+            table_list.append(sublist_list)
+        x_values_list.append(table_list)
+        # headers list
+        table_list = []
+        for col in df_headers.columns:
+            sublist_list = df_headers[col].tolist()
+            table_list.append(sublist_list)
+        x_headers_list.append(table_list)
+        # y list
+        y_list.append(y)
+
+    return x_values_list, x_headers_list, y_list
+
+
+def lazy_loading(data_list: List, batch_size: int) -> Iterator[List]:
+    """
+    Lazy loading for data in batches.
+
+    :param List data_list: List of data to be loaded lazily.
+    :param int batch_size: Size of batch.
+    """
+    for i in range(0, len(data_list), batch_size):
+        yield data_list[i : i + batch_size]
+
+
+def get_top_training_cluster_averaged(
+    embeddings: List[torch.tensor], num: int
+) -> torch.Tensor:
+    """
+    Computes the clutser-averaged top training embeddings using k-means clustering.
+
+    :param List[torch.tensor] embeddings: List of embedding tensors to cluster.
+    :param int num: Number of clusters to be created using k-means.
+    :return torch.Tensor: A tensor representing the
+        average of embeddings in the most common cluster.
+    """
+    flattened_embeddings = [embedding.tolist() for embedding in embeddings]
```

Comment:
I dont think this is flattened, its just converting the list of `np.array`s to a list of lists, right?